### PR TITLE
Set ipBUFFER_PADDING to 14 bytes by default on 64 bit targets v2

### PR DESCRIPTION
**Pull request by htibosch**
Description
-----------
While looking at old un-merged PRs I saw [PR #547](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/547): "Set ipBUFFER_PADDING to 14 bytes by default on 64 bit targets".

Mentioned PR had the following description:
If `ipconfigBUFFER_PADDING` is not defined in FreeRTOSIPConfig.h, builds for 64 bit targets fail. Add a default value of 14 bytes `( 12U + ipconfigPACKET_FILLER_SIZE )` for `ipBUFFER_PADDING` to satisfy the alignment check in FreeRTOS_IP_Utils.c.

The change in FreeRTOS_IP.h:
~~~diff
     /* Use setting from FreeRTOS if defined and non-zero */
     #if defined( ipconfigBUFFER_PADDING ) && ( ipconfigBUFFER_PADDING != 0 )
         #define ipBUFFER_PADDING    ipconfigBUFFER_PADDING
+    #elif ( UINTPTR_MAX > 0xFFFFFFFF )
+        #define ipBUFFER_PADDING    ( 12U + ipconfigPACKET_FILLER_SIZE )
     #else
         #define ipBUFFER_PADDING    ( 8U + ipconfigPACKET_FILLER_SIZE )
     #endif
~~~

The PR was reviewed, approved, and then somehow forgotten.

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [ ] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: source/FreeRTOS_IP_Utils.c [hunk 1]
```diff
@@ -983,18 +983,22 @@ void vPreCheckConfigs( void )
...
+             * This is a 64-bit platform, make sure there is enough space in
+             * pucEthernetBuffer to store a pointer.
+             */
+            configASSERT( ipBUFFER_PADDING >= 14U );
+        #else
+            /* This is a 32-bit platform. */
+            configASSERT( ipBUFFER_PADDING >= 10U );
+        #endif /* UINTPTR_MAX > 0xFFFFFFFFU */
+
+        /* And it must have this strange alignment: */
```
Comment (tony-josi-aws) [inline source/FreeRTOS_IP_Utils.c:1000]: Can we add that the purpose of this assert is to make sure that the contents of the network packets after the ethernet header lands in a 32 bit aligned address as a comment?

I think, changing the assert to: `configASSERT( ( ( ( ipBUFFER_PADDING ) + ( ipSIZE_OF_ETH_HEADER ) ) % 4U ) == 0U );` would also make it more clear to the reader.

File: source/FreeRTOS_IP_Utils.c [hunk 2]
```diff
@@ -983,18 +983,22 @@ void vPreCheckConfigs( void )
...
+        configASSERT( ( ( ( ipBUFFER_PADDING ) + 2U ) % 4U ) == 0U );
```
Comment (tony-josi-aws) [inline source/FreeRTOS_IP_Utils.c:1001]: I think, changing the assert to: `configASSERT( ( ( ( ipBUFFER_PADDING ) + ( ipSIZE_OF_ETH_HEADER ) ) % 4U ) == 0U );` would also make it more clear to the reader.

File: source/include/FreeRTOS_IP.h [hunk 3]
```diff
@@ -97,22 +97,29 @@ extern uint32_t ulApplicationGetNextSequenceNumber( uint32_t ulSourceAddress,
...
- *  uchar16_t type;
- *  << IP-header >>     // word-aligned
+ *
+ * +---------+-----------+---------+
+ * | Offset  | Alignment | Length  |
+ * | 32 | 64 | 32  | 64  | 32 | 64 |
+ * |----|----|-----|-----|----|----|
+ * | 0  | 0  | 4   | 8   | 4  | 8  | uchar_8 * pointer;     // Points to the 'NetworkBufferDescriptor_t'.
+ * | 4  | 8  | 4   | 8   | 2  | 2  | uchar_8   filler[];    // To give the +2 byte offset.
+ * | 6  | 10 | 4+2 | 8+2 | 6  | 6  | uchar_8   dest_mac[6]; // Destination address.
```
Comment (tony-josi-aws) [inline source/include/FreeRTOS_IP.h:110]: In the case of 32 bit platforms the `ipBUFFER_PADDING` is `( 8U + ipconfigPACKET_FILLER_SIZE )` by default. And `ipconfigPACKET_FILLER_SIZE` is 2 by default. In that case shouldn't the `dest_mac` start at an offset of 10 bytes from the start of the network buffer. Similarly 14 bytes for 64 bit platforms?

File: source/FreeRTOS_IP_Utils.c (hunk 2 above)
Comment (moninom1) [reply source/FreeRTOS_IP_Utils.c:1001]: On the same line , can we convert all the numbers used above for comparison, to a descriptive MACROs to be more clear.

File: source/include/FreeRTOS_IP.h (hunk 3 above)
Comment (htibosch) [reply source/include/FreeRTOS_IP.h]: There were indeed some typos in the table in the comments. Hopefully all correct now.

File: source/FreeRTOS_IP_Utils.c (hunk 1 above)
Comment (htibosch) [reply source/FreeRTOS_IP_Utils.c]: We have explained it already in other locations. I added a short comment to the asserts:
~~~diff
+    /* Check if ipBUFFER_PADDING has a minimum size, depending on the platform.
+     * See FreeRTOS_IP.h for more details. */
~~~

File: source/FreeRTOS_IP_Utils.c (hunk 2 above)
Comment (htibosch) [reply source/FreeRTOS_IP_Utils.c]: Good suggestion to calculate the sum of these two macros :
~~~diff
+configASSERT( ( ( ( ipSIZE_OF_ETH_HEADER ) + ( ipBUFFER_PADDING ) ) % 4U ) == 0U );
~~~
and make sure the result is a multiple of 4 bytes.

**Review by tony-josi-aws — APPROVED** [review]

**Review by aggarg — APPROVED** [review]

**Pull request merged**